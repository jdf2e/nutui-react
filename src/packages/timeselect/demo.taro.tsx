import React, { useState } from 'react'
import { useTranslate } from '@/sites/assets/locale/taro'
import { TimeSelect, Cell } from '@/packages/nutui.react.taro'

export interface TimeType {
  key?: string | number
  list: string[]
}

const TimeSelectDemo = () => {
  const [translated] = useTranslate<any>({
    'zh-CN': {
      text1: '您选择了',
      text2: '请选择配送时间',
      text3: '配送时间',
      text4: '5月20日(今天)',
      text5: '5月21日(星期三)',
    },
    'zh-TW': {
      text1: '您选择了',
      text2: '请选择配送时间',
      text3: '配送时间',
      text4: '5月20日(今天)',
      text5: '5月21日(星期三)',
    },
    'en-US': {
      text1: 'selected',
      text2: 'delivery time',
      text3: 'Pickup time',
      text4: 'May 20 (Today)',
      text5: 'May 21 (Wednesday)',
    },
  })
  const [visible1, SetVisible1] = useState(false)
  const currentKey = 0

  const dates = [
    {
      'pannel-key': '0',
      date: translated.text4,
    },
    {
      'pannel-key': '1',
      date: translated.text5,
    },
  ]
  const times = [
    {
      key: '0',
      list: ['9:00-10:00', '10:00-11:00', '11:00-12:00'],
    },
    {
      key: '1',
      list: ['9:00-10:00', '10:00-11:00'],
    },
  ]

  const handleClick = () => {
    SetVisible1(true)
  }
  // 点击弹层 X 或者弹层外区域触发事件
  const handleSelect = (selectTimeData: TimeType[]) => {
    SetVisible1(false)
    console.log(`您选择了: ${JSON.stringify(selectTimeData)}`)
    // Toast.text(`您选择了: ${JSON.stringify(selectTimeData)}`)
  }
  // 选择日期触发回调事件
  const handlePannelChange = (
    pannelKey: string | number,
    selectTimeData: TimeType[]
  ) => {
    console.log('pannelKey, selectTimeData: ', pannelKey, selectTimeData)
  }
  // 选择配送时间触发回调事件
  const handleTimeChange = (time: string, selectTimeData: TimeType[]) => {
    console.log('time, selectTimeData: ', time, selectTimeData)
  }
  return (
    <>
      <div className="demo">
        <Cell title={translated.text2} click={handleClick} />
        <TimeSelect
          visible={visible1}
          height="50%"
          title={translated.text3}
          multiple
          currentKey={currentKey}
          dates={dates}
          times={times}
          onSelect={handleSelect}
          onPannelChange={handlePannelChange}
          onTimeChange={handleTimeChange}
        />
      </div>
    </>
  )
}

export default TimeSelectDemo
