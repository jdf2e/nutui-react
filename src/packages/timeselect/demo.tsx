import React, { useState } from 'react'
import TimeSelect, { DateType, TimeType } from '@/packages/timeselect'
import Cell from '@/packages/cell'
import Toast from '@/packages/toast'
import { useTranslate } from '@/sites/assets/locale'

const TimeSelectDemo = () => {
  const [translated] = useTranslate<any>({
    'zh-CN': {
      text1: '您选择了',
      text2: '请选择配送时间',
      text3: '配送时间',
      text4: '5月20日(今天)',
      text5: '5月21日(星期三)',
      basic: '基础用法',
      optionKey: '自定义数据 key',
      multiple: '支持多选',
      custom: '自定义使用场景',
      city: '请选择所在城市',
      zhejiang: '浙江',
      hangzhou: '杭州',
      ningbo: '宁波',
      jiangsu: '江苏',
      nanjing: '南京',
      suzhou: '苏州',
      yangzhou: '扬州',
    },
    'zh-TW': {
      text1: '您选择了',
      text2: '请选择配送时间',
      text3: '配送时间',
      text4: '5月20日(今天)',
      text5: '5月21日(星期三)',
      basic: '基础用法',
      optionKey: '自定义数据 key',
      multiple: '支持多选',
      custom: '自定义使用场景',
      city: '请选择所在城市',
      zhejiang: '浙江',
      hangzhou: '杭州',
      ningbo: '宁波',
      jiangsu: '江苏',
      nanjing: '南京',
      suzhou: '苏州',
      yangzhou: '扬州',
    },
    'en-US': {
      text1: 'selected',
      text2: 'delivery time',
      text3: 'Pickup time',
      text4: 'May 20 (Today)',
      text5: 'May 21 (Wednesday)',
      basic: 'Basic Usage',
      optionKey: 'Custom optionKey',
      multiple: 'Multiple Mode',
      custom: 'Custom Usage',
      city: 'Select your city',
      zhejiang: 'ZheJiang',
      hangzhou: 'HangZhou',
      ningbo: 'NingBo',
      jiangsu: 'JiangSu',
      nanjing: 'NanJing',
      suzhou: 'SuZhou',
      yangzhou: 'YangZhou',
    },
  })

  const [visible1, setVisible1] = useState(false)
  const [visible2, setVisible2] = useState(false)
  const [visible3, setVisible3] = useState(false)
  const [visible4, setVisible4] = useState(false)
  const options1 = [
    {
      value: '20230520',
      text: translated.text4,
      children: [
        { value: '09', text: '09:00-10:00' },
        { value: '10', text: '10:00-11:00' },
        { value: '11', text: '11:00-12:00' },
      ],
    },
    {
      value: '20230521',
      text: translated.text5,
      children: [
        { value: '09', text: '09:00-10:00' },
        { value: '10', text: '10:00-11:00' },
      ],
    },
  ]

  const optionKey = {
    valueKey: 'value1',
    textKey: 'text1',
    childrenKey: 'children1',
  }
  const options2 = [
    {
      value1: '20230520',
      text1: translated.text4,
      children1: [
        { value1: '09', text1: '09:00-10:00' },
        { value1: '10', text1: '10:00-11:00' },
        { value1: '11', text1: '11:00-12:00' },
      ],
    },
    {
      value1: '20230521',
      text1: translated.text5,
      children1: [
        { value1: '09', text1: '09:00-10:00' },
        { value1: '10', text1: '10:00-11:00' },
      ],
    },
  ]
  const defaultValue2 = [
    {
      value1: '20230521',
      children1: [{ value1: '10' }],
    },
  ]
  const options4 = [
    {
      value: 'zhejiang',
      text: translated.zhejiang,
      children: [
        { value: 'hangzhou', text: translated.hangzhou },
        { value: 'ningbo', text: translated.ningbo },
      ],
    },
    {
      value: 'jiangsu',
      text: translated.jiangsu,
      children: [
        { value: 'nanjing', text: translated.nanjing },
        { value: 'suzhou', text: translated.suzhou },
        { value: 'yangzhou', text: translated.yangzhou },
      ],
    },
  ]

  const handleClick1 = () => {
    setVisible1(true)
  }
  const handleClick2 = () => {
    setVisible2(true)
  }
  const handleClick3 = () => {
    setVisible3(true)
  }
  const handleClick4 = () => {
    setVisible4(true)
  }
  const handleSelect1 = (value: DateType[]) => {
    setVisible1(false)
    Toast.show(`${translated.text1}: ${JSON.stringify(value)}`)
  }
  const handleSelect2 = (value: DateType[]) => {
    setVisible2(false)
    Toast.show(`${translated.text1}: ${JSON.stringify(value)}`)
  }
  const handleSelect3 = (value: DateType[]) => {
    setVisible3(false)
    Toast.show(`${translated.text1}: ${JSON.stringify(value)}`)
  }
  const handleSelect4 = (value: DateType[]) => {
    setVisible4(false)
    Toast.show(`${translated.text1}: ${JSON.stringify(value)}`)
  }
  // 选择日期触发回调事件
  const handleDateChange = (date: DateType, value: DateType[]) => {
    console.log(date, value)
  }
  // 选择配送时间触发回调事件
  const handleTimeChange = (time: TimeType, value: DateType[]) => {
    console.log(time, value)
  }
  return (
    <>
      <div className="demo">
        <h2>{translated.basic}</h2>
        <Cell title={translated.text2} onClick={handleClick1} />
        <TimeSelect
          visible={visible1}
          options={options1}
          style={{
            height: '30%',
          }}
          onSelect={handleSelect1}
          onDateChange={handleDateChange}
          onTimeChange={handleTimeChange}
        />

        <h2>{translated.optionKey}</h2>
        <Cell title={translated.text2} onClick={handleClick2} />
        <TimeSelect
          options={options2}
          optionKey={optionKey}
          defaultValue={defaultValue2}
          visible={visible2}
          style={{
            height: '30%',
          }}
          onSelect={handleSelect2}
          onDateChange={handleDateChange}
          onTimeChange={handleTimeChange}
        />

        <h2>{translated.multiple}</h2>
        <Cell title={translated.text2} onClick={handleClick3} />
        <TimeSelect
          options={options1}
          visible={visible3}
          style={{
            height: '30%',
          }}
          multiple
          onSelect={handleSelect3}
          onDateChange={handleDateChange}
          onTimeChange={handleTimeChange}
        />

        <h2>{translated.custom}</h2>
        <Cell title={translated.city} onClick={handleClick4} />
        <TimeSelect
          options={options4}
          visible={visible4}
          style={{
            height: '30%',
          }}
          title={translated.city}
          onSelect={handleSelect4}
          onDateChange={handleDateChange}
          onTimeChange={handleTimeChange}
        />
      </div>
    </>
  )
}

export default TimeSelectDemo
