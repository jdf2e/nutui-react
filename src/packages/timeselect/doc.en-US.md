# TimeSelect 配送时间

### Intro

For delivery time selection

### Install

``` javascript
import { TimeSelect } from '@nutui/nutui-react';
```

### Basic Usage

:::demo

```tsx
import React, { useState } from 'react'
import { TimeSelect, Cell, Toast } from '@nutui/nutui-react'

const TimeSelectDemo = () => {
  const [visible1, SetVisible1] = useState(false)
  const currentKey = 0

  const dates = [
    {
      'pannel-key': '0',
      date: 'May 20 (Today)',
    },
    {
      'pannel-key': '1',
      date: 'May 21 (Wednesday)',
    },
  ]
  const times = [
    {
      key: '0',
      list: ['9:00-10:00', '10:00-11:00', '11:00-12:00'],
    },
    {
      key: '1',
      list: ['9:00-10:00', '10:00-11:00'],
    },
  ]

  const handleClick = () => {
    SetVisible1(true)
  }
  const handleSelect = (selectTimeData) => {
    SetVisible1(false)
    Toast.text(`selected: ${JSON.stringify(selectTimeData)}`)
  }
  const handlePannelChange = (pannelKey, selectTimeData) => {
    console.log('pannelKey, selectTimeData: ', pannelKey, selectTimeData)
  }
  const handleTimeChange = (time, selectTimeData) => {
    console.log('time, selectTimeData: ', time, selectTimeData)
  }
  return (
    <>
      <div className="demo">
        <h2>Demo</h2>
        <Cell title="delivery time" click={handleClick} />
        <TimeSelect
          visible={visible1}
          height="50%"
          title="Pickup time"
          multiple
          currentKey={currentKey}
          dates={dates}
          times={times}
          onSelect={handleSelect}
          onPannelChange={handlePannelChange}
          onTimeChange={handleTimeChange}
        />
      </div>
    </>
  )
}

export default TimeSelectDemo
```

:::

## API

### TimeSelect Prop

| Props                   | Description                 | required | Type    | Default |
|------------------------|-----------------------------|----------|------------|---------|
| visible                 | display the popup           | true     | Boolean  | `false`|
| height                 | height of the popup         | false        | String  | `20%`|
| title                 | popup title                 | false                | String  | `Pickup time`|
| multiple              | Select multiple datetimes   | false                | String  | `false`|
| currentKey           | Uniquely identifies         | false                | String、Number  | `0` |
| dates            | Select data for date panel  | true     | []        | ``       |
| times            | Select data for time panel  | true     | []        | ``       |

### dates

| Props                   | Description                                                             | Type    | Default |
|------------------------|----------------------------------------------------------------|---------|------|
| date                 | display name                                            | String  | ``|
| pannel-key           | Unique identifier, together with currentKey to identify the currently selected day            | Number、String  | `0`|

### times

| Props                   | Description                                                                                       | Type    | Default |
|------------------------|---------------------------------------------------------------------------------------------------|---------|------|
| key                 | Unique identifier, together with pannel-key and currentKey to identify the currently selected day | Array  | `[]`|
| list                 | Optional time list                                                                                | Array  | `[]`|

### TimeSelect Event

| Event                  | Description                                                                                                       | required | callback args                         |
|------------------------|-------------------------------------------------------------------------------------------------------------------|----------|---------------------------------------|
| onSelect`v1.3.8`       | Callback after closing the mask                                                                                   | true     | `list: []`                            |
| onPannelChange`v1.3.8` | Click the date callback in the left column, and internally update the data through setCurrentKey, setCurrentTime  | false    | `pannelKey: string 、number, list: []` |
| onTimeChange`v1.3.8`         | Click the time callback, and internally update the data through setCurrentKey, setCurrentTime                                                                     | false                | `time: string, list: []`              |
