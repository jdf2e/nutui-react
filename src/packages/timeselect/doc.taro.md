# TimeSelect 配送时间

## 介绍

用于配送时间选择

## 安装

```tsx
import { TimeSelect } from '@nutui/nutui-react-taro';
```

## 代码演示

### 基础用法

:::demo

```tsx
import React, { useState } from 'react'
import { TimeSelect, Cell } from '@nutui/nutui-react-taro'

const Demo = () => {
  const [visible, SetVisible] = useState(false)
  const options = [
    {
      value: '20230520',
      text: '5月20日(今天)',
      children: [
        { value: '09', text: '09:00-10:00' },
        { value: '10', text: '10:00-11:00' },
        { value: '11', text: '11:00-12:00' },
      ],
    },
    {
      value: '20230521',
      text: '5月21日(星期三)',
      children: [
        { value: '09', text: '09:00-10:00' },
        { value: '10', text: '10:00-11:00' },
      ],
    },
  ]
  const handleClick = () => {
    setVisible(true)
  }
  const handleSelect = (value) => {
    setVisible(false)
    console.log(`您选择了: ${JSON.stringify(value)}`)
  }
  const handleDateChange = (date, value) => {
    console.log(date, value)
  }
  const handleTimeChange = (time, value) => {
    console.log(time, value)
  }
  return (
    <>
      <Cell title="请选择配送时间" onClick={handleClick} />
      <TimeSelect
        visible={visible}
        options={options}
        style={{
          height: '30%',
        }}
        onSelect={handleSelect}
        onDateChange={handleDateChange}
        onTimeChange={handleTimeChange}
      />
    </>
  )
}

export default Demo
```

:::

### 自定义数据 key

:::demo

```tsx
import React, { useState } from 'react'
import { TimeSelect, Cell } from '@nutui/nutui-react-taro'

const Demo = () => {
  const [visible, SetVisible] = useState(false)
  const optionKey = {
    valueKey: 'value1',
    textKey: 'text1',
    childrenKey: 'children1',
  }
  const options = [
    {
      value1: '20230520',
      text1: '5月20日(今天)',
      children1: [
        { value1: '09', text1: '09:00-10:00' },
        { value1: '10', text1: '10:00-11:00' },
        { value1: '11', text1: '11:00-12:00' },
      ],
    },
    {
      value1: '20230521',
      text1: '5月21日(星期三)',
      children1: [
        { value1: '09', text1: '09:00-10:00' },
        { value1: '10', text1: '10:00-11:00' },
      ],
    },
  ]
  const defaultValue = [
    {
      value1: '20230521',
      children1: [{ value1: '10' }],
    },
  ]
  const handleClick = () => {
    setVisible(true)
  }
  const handleSelect = (value) => {
    setVisible(false)
    console.log(`您选择了: ${JSON.stringify(value)}`)
  }
  const handleDateChange = (date, value) => {
    console.log(date, value)
  }
  const handleTimeChange = (time, value) => {
    console.log(time, value)
  }
  return (
    <>
      <Cell title="请选择配送时间" onClick={handleClick} />
      <TimeSelect
        options={options}
        optionKey={optionKey}
        defaultValue={defaultValue}
        visible={visible}
        style={{
          height: '30%',
        }}
        onSelect={handleSelect}
        onDateChange={handleDateChange}
        onTimeChange={handleTimeChange}
      />
    </>
  )
}

export default Demo
```

:::

### 支持多选

:::demo

```tsx
import React, { useState } from 'react'
import { TimeSelect, Cell } from '@nutui/nutui-react-taro'

const Demo = () => {
  const [visible, SetVisible] = useState(false)
  const options = [
    {
      value: '20230520',
      text: '5月20日(今天)',
      children: [
        { value: '09', text: '09:00-10:00' },
        { value: '10', text: '10:00-11:00' },
        { value: '11', text: '11:00-12:00' },
      ],
    },
    {
      value: '20230521',
      text: '5月21日(星期三)',
      children: [
        { value: '09', text: '09:00-10:00' },
        { value: '10', text: '10:00-11:00' },
      ],
    },
  ]
  const handleClick = () => {
    setVisible(true)
  }
  const handleSelect = (value) => {
    setVisible(false)
    console.log(`您选择了: ${JSON.stringify(value)}`)
  }
  const handleDateChange = (date, value) => {
    console.log(date, value)
  }
  const handleTimeChange = (time, value) => {
    console.log(time, value)
  }
  return (
    <>
      <Cell title="请选择配送时间" onClick={handleClick} />
      <TimeSelect
        visible={visible}
        options={options}
        style={{
          height: '30%',
        }}
        multiple
        onSelect={handleSelect}
        onDateChange={handleDateChange}
        onTimeChange={handleTimeChange}
      />
    </>
  )
}

export default Demo
```

:::

### 自定义使用场景

:::demo

```tsx
import React, { useState } from 'react'
import { TimeSelect, Cell } from '@nutui/nutui-react-taro'

const Demo = () => {
  const [visible, SetVisible] = useState(false)
  const options = [
    {
      value: 'zhejiang',
      text: '浙江',
      children: [
        { value: 'hangzhou', text: '杭州' },
        { value: 'ningbo', text: '宁波' },
      ],
    },
    {
      value: 'jiangsu',
      text: '江苏',
      children: [
        { value: 'nanjing', text: '南京' },
        { value: 'suzhou', text: '苏州' },
        { value: 'yangzhou', text: '扬州' },
      ],
    },
  ]
  const handleClick = () => {
    setVisible(true)
  }
  const handleSelect = (value) => {
    setVisible(false)
    console.log(`您选择了: ${JSON.stringify(value)}`)
  }
  const handleDateChange = (date, value) => {
    console.log(date, value)
  }
  const handleTimeChange = (time, value) => {
    console.log(time, value)
  }
  return (
    <>
      <Cell title="请选择所在城市" onClick={handleClick} />
      <TimeSelect
        visible={visible}
        options={options}
        style={{
          height: '30%',
        }}
        title="请选择所在城市"
        onSelect={handleSelect}
        onDateChange={handleDateChange}
        onTimeChange={handleTimeChange}
      />
    </>
  )
}

export default Demo
```

:::

## TimeSelect

### Props

| 属性 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| visible | 是否显示弹层 | `boolean` | `false` |
| title | 弹层标题 | `ReactNode` | `取件时间` |
| multiple | 是否支持多选 | `boolean` | `false` |
| defaultValue | 默认选中的值，非受控 | `DateType[]` | `-` |
| options | 数据 | `DateType[]` | `-` |
| optionKey | 配置数据中的关键字, `valueKey`, `textKey`, `childrenKey` | `{valueKey: 'value', textKey: 'text', childrenKey: 'children'}` | `-` |
| onSelect | 关闭遮罩之后的回调 | `(value: DateType[]) => void` | `-` |
| onDateChange | 点击左栏时的回调 | `(date: DateType, value: DateType[]) => void` | `-` |
| onTimeChange | 点击右侧选项时的回调 | `(time: TimeType, value: DateType[]) => void` | `-` |

### DateType

以下字段为默认值，会被 optionKey 的配置替换。

| 属性 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| value | 唯一标识符， 必填 | `string` | `-` |
| text | 左侧显示的文本， 必填 | `string` | `-` |
| children | 对应右侧的选项列表， 必填 | `TimeType[]` | `-` |

### TimeType

以下字段为默认值，会被 optionKey 的配置替换。

| 属性 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| value | 唯一标识符， 必填 | `string` | `-` |
| text | 右侧显示的选项内容， 必填 | `string` | `-` |

## 主题定制

### 样式变量

组件提供了下列 CSS 变量，可用于自定义样式，使用方法请参考 [ConfigProvider 组件](#/zh-CN/component/configprovider)。

| 名称 | 说明 | 默认值 |
| --- | --- | --- |
| \--nutui-timeselect-title-height | title 高度 | `50px` |
| \--nutui-timeselect-title-line-height | title 行高 | `50px` |
| \--nutui-timeselect-title-color | title 字体颜色 | `$title-color` |
| \--nutui-timeselect-title-font-size | title 字体大小 | `$font-size-2` |
| \--nutui-timeselect-title-background | title 背景 | `$background-color3` |
| \--nutui-timeselect-date-width | date 宽度 | `140px` |
| \--nutui-timeselect-date-height | date 高度 | `40px` |
| \--nutui-timeselect-date-line-height | date 行高 | `40px` |
| \--nutui-timeselect-date-color | date 字体颜色 | `$text-color` |
| \--nutui-timeselect-date-active-color | date 激活字体颜色 | `$title-color` |
| \--nutui-timeselect-date-font-size | date 字体大小 | `$font-size-2` |
| \--nutui-timeselect-time-width | time 宽度 | `100px` |
| \--nutui-timeselect-time-height | time 高度 | `50px` |
| \--nutui-timeselect-time-line-height | time 行高 | `50px` |
| \--nutui-timeselect-time-font-size | time 字体大小 | `14px` |
| \--nutui-timeselect-time-margin | time 外边距 | `0 10px 10px 0` |
| \--nutui-timeselect-time-color | time 字体颜色 | `$title-color` |
| \--nutui-timeselect-time-active-color | time 激活字体颜色 | `$primary-color` |
| \--nutui-timeselect-time-background | time 背景 | `$background-color` |
| \--nutui-timeselect-time-active-background | time 激活背景 | `$background-color2` |