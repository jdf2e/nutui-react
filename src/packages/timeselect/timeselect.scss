@import '../popup/popup.scss';
@import '../timedetail/timedetail.scss';

.nut-timeselect {
  background-color: $color-background-overlay;
  display: flex;
  flex-direction: column;
  height: calc(100% - 50px);

  &__content {
    display: flex;
    flex: 1;
  }

  &__content-left {
    width: $timeselect-date-width;
    min-width: $timeselect-date-width;
    height: 100%;
    overflow: auto;
    background: $color-background-sunken;
  }
}

.nut-timepannel {
  padding: 0 16px;
  height: $timeselect-date-height;
  line-height: $timeselect-date-height;
  text-align: left;
  color: $color-text;
  font-size: $font-size;

  &.active {
    background: $color-background-overlay;
    color: $timeselect-date-active-color;
    font-weight: $font-weight-bold;
  }
}
