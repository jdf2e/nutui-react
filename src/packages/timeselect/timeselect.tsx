import React, { FunctionComponent, ReactNode, useState } from 'react'
import classNames from 'classnames'
import Popup from '@/packages/popup'
import TimePannel from '@/packages/timepannel'
import TimeDetail from '@/packages/timedetail'
import { TimeType } from '@/packages/timedetail/timedetail'
import { BasicComponent, ComponentDefaults } from '@/utils/typings'

export interface DateType {
  paneKey?: string | number
  date: string
}
export interface TimeSelectProps extends BasicComponent {
  visible: boolean
  multiple?: boolean
  title?: ReactNode
  currentKey: string | number
  currentTime: TimeType[]
  dates: DateType[]
  times: TimeType[]
  select?: (selectTimeData: TimeType[]) => void
  pannelChange?: (
    pannelKey: string | number,
    selectTimeData: TimeType[]
  ) => void
  timeChange?: (time: string, selectTimeData: TimeType[]) => void
  onSelect?: (selectTimeData: TimeType[]) => void
  onDateChange?: (
    pannelKey: string | number,
    selectTimeData: TimeType[]
  ) => void
  onTimeChange?: (time: string, selectTimeData: TimeType[]) => void
}
const defaultProps = {
  ...ComponentDefaults,
  visible: false,
  multiple: false,
  currentKey: 0,
  currentTime: [],
  title: '取件时间',
  dates: [],
  times: [],
} as TimeSelectProps
export const TimeSelect: FunctionComponent<Partial<TimeSelectProps>> = (
  props
) => {
  const {
    visible,
    className,
    style,
    title,
    currentKey,
    currentTime,
    dates,
    times,
    multiple,
    select,
    pannelChange,
    timeChange,
    onSelect,
    onDateChange,
    onTimeChange,
    ...rest
  } = {
    ...defaultProps,
    ...props,
  }
  const [activeKey, setActiveKey] = useState<string | number>(currentKey)
  const [activeTime, setActiveTime] = useState<TimeType[]>(
    currentTime.length
      ? currentTime
      : [
          {
            key: currentKey,
            list: [],
          },
        ]
  )
  const classPrefix = 'nut-timeselect'

  // popup 的关闭回调, 执行 select 函数更改外部 visible
  const closeFun = () => {
    if (onSelect) {
      onSelect(activeTime)
    } else if (select) {
      select(activeTime)
    }
  }
  // 选择配送时间回调
  const handleSelectTime = (time: string) => {
    let curTimeData = {} as TimeType
    let curIndex: string | number = -1
    for (let i = 0; i < activeTime.length; i++) {
      if (String(activeTime[i].key) === String(activeKey)) {
        curTimeData = activeTime[i]
        curIndex = i
        break
      }
    }
    const curTimeIndex = curTimeData.list.findIndex(
      (item: string) => String(item) === String(time)
    )
    if (curTimeIndex === -1) {
      curTimeData.list.push(time)
    } else {
      curTimeData.list.splice(curTimeIndex, 1)
    }
    const resultTimeData = [...activeTime]
    resultTimeData.splice(curIndex, 1, curTimeData)
    setActiveTime(resultTimeData)
    if (onTimeChange) {
      onTimeChange(time, resultTimeData)
    } else if (timeChange) {
      timeChange(time, resultTimeData)
    }
  }
  // 选择日期的回调
  const handleChange = (pannelKey: string | number) => {
    const resultTimeData: TimeType[] = [...activeTime]
    if (String(pannelKey) !== String(activeKey)) {
      setActiveKey?.(pannelKey)
      if (multiple) {
        const curTimeDataIndex = activeTime.findIndex(
          (item: TimeType) => String(item.key) === String(pannelKey)
        )
        if (curTimeDataIndex === -1) {
          resultTimeData.push({
            key: pannelKey,
            list: [],
          } as TimeType)
          setActiveTime(resultTimeData)
        }
      } else {
        setActiveTime([
          {
            key: pannelKey,
            list: [],
          } as TimeType,
        ])
      }
    }
    if (onDateChange) {
      onDateChange(pannelKey, resultTimeData)
    } else if (pannelChange) {
      pannelChange(pannelKey, resultTimeData)
    }
  }
  // 选中的日期增加 active 类名
  const getTimePannelClass = (dataItem: DateType) => {
    if (String(dataItem['paneKey']) === String(activeKey)) {
      return 'nut-timepannel-active'
    }
    return ''
  }
  return (
    <Popup
      closeable
      round
      visible={visible}
      position="bottom"
      style={{
        width: '100%',
        height: '20%',
        ...style,
      }}
      onClose={closeFun}
      {...rest}
    >
      <div className={classNames(classPrefix, className)}>
        <div className={`${classPrefix}__title`}>{title}</div>
        <div className={`${classPrefix}__content`}>
          <div className={`${classPrefix}__content-left`}>
            {dates.map((dataItem: DateType, index: number) => (
              <TimePannel
                date={dataItem.date}
                className={getTimePannelClass(dataItem)}
                key={String(dataItem['paneKey'] || index)}
                curKey={String(dataItem['paneKey'] || index)}
                onChange={handleChange}
              />
            ))}
          </div>
          <TimeDetail
            times={times}
            currentKey={String(activeKey)}
            currentTime={activeTime}
            onSelect={handleSelectTime}
          />
        </div>
      </div>
    </Popup>
  )
}

TimeSelect.defaultProps = defaultProps
TimeSelect.displayName = 'NutTimeSelect'
