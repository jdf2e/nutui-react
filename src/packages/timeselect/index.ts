import { TimeSelect } from './timeselect'

export default TimeSelect
