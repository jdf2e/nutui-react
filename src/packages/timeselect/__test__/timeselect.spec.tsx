import * as React from 'react'

import { render, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import ReactTestUtils from 'react-dom/test-utils'

import { TimeSelect } from '../timeselect'

const state = {
  visible1: true,
  height: '50%',
  title: '取件时间 1',
  currentKey: 0,
  dates: [
    {
      'pannel-key': '0',
      date: '5月20日(今天)',
    },
    {
      'pannel-key': '1',
      date: '5月21日(星期三)',
    },
  ],
  times: [
    {
      key: '0',
      list: ['9:00-10:00', '10:00-11:00', '11:00-12:00'],
    },
    {
      key: '1',
      list: ['9:00-10:00', '10:00-11:00'],
    },
  ],
}
test('timeselect props test', async () => {
  const App = () => {
    return (
      <TimeSelect
        visible={state.visible1}
        height={state.height}
        title={state.title}
        multiple
        currentKey={state.currentKey}
        dates={state.dates}
        times={state.times}
      />
    )
  }
  const { container } = render(<App />)

  await waitFor(
    () => {
      const titleDoms = container.querySelectorAll('.nut-timeselect__title')
      expect(titleDoms[0].innerHTML).toBe(state.title)
      container.querySelectorAll('.nut-timepannel').forEach((item, index) => {
        expect(item.innerHTML).toBe(state.dates[index].date)
      })
    },
    { timeout: 2000 }
  )

  await waitFor(
    () => {
      expect(container.querySelector('.nut-timedetail')).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    },
    { timeout: 2000 }
  )
})

test('timeselect event test', async () => {
  const App = () => {
    return (
      <TimeSelect
        visible={state.visible1}
        height={state.height}
        title={state.title}
        multiple
        currentKey={state.currentKey}
        dates={state.dates}
        times={state.times}
      />
    )
  }
  const { container } = render(<App />)
  await waitFor(
    () => {
      container
        .querySelectorAll('.nut-timedetail__item')
        .forEach((item, index) => {
          ReactTestUtils.Simulate.click(item)
          expect(item).toHaveClass('nut-timedetail__item-active')
        })
    },
    { timeout: 2000 }
  )
})
