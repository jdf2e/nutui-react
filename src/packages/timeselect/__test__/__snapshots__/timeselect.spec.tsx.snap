// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeselect props test 1`] = `"<div><div class=\\"nut-overlay-fade-enter-active nut-overlay\\" style=\\"z-index: 1000; animation-duration: 0.3s;\\"></div><div style=\\"width: 100%; height: 20%; z-index: 1000; animation-duration: 0.3s;\\" class=\\"round nut-popup nut-popup-bottom \\"><div class=\\"nut-popup__close-icon nut-popup__close-icon--top-right\\"><svg class=\\"nut-icon nut-icon-Close\\" style=\\"width: 12px; height: 12px;\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 1026 1024\\" aria-labelledby=\\"Close\\" role=\\"presentation\\"><path d=\\"M981.577 1024c-11.703 0-23.406-2.926-32.183-11.703L13.166 76.07c-14.629-17.555-14.629-46.812 0-64.366 17.554-14.629 46.811-14.629 64.365 0L1013.76 947.93c17.554 17.555 17.554 43.886 0 61.44-8.777 11.703-20.48 14.629-32.183 14.629zm-936.228 0c-11.703 0-23.406-2.926-32.183-11.703-17.555-17.554-17.555-43.886 0-61.44L949.394 14.63c17.555-17.555 43.886-17.555 61.44 0 17.555 17.554 17.555 43.885 0 61.44L74.606 1012.297C68.754 1021.074 57.05 1024 45.349 1024z\\" fill=\\"currentColor\\" fill-opacity=\\"0.9\\"></path></svg></div><div class=\\"nut-timeselect\\"><div class=\\"nut-timeselect__title\\">取件时间</div><div class=\\"nut-timeselect__content\\"><div class=\\"nut-timeselect__content-left\\"><div class=\\"nut-timepannel active\\">520</div><div class=\\"nut-timepannel\\">521</div></div><div class=\\"nut-timedetail\\"><span class=\\"nut-timedetail__item\\">09:00-10:00</span><span class=\\"nut-timedetail__item\\">10:00-11:00</span><span class=\\"nut-timedetail__item\\">11:00-12:00</span></div></div></div></div></div>"`;
